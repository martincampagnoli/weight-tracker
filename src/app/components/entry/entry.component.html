<mat-card appearance="outlined" class="entry-card">
  <mat-card-header class="entry-header">
    <mat-card-title class="entry-title">
      <span>{{ entry().weight }} kgs</span>
    </mat-card-title>
    <mat-card-subtitle class="entry-subtitle">
      <span>{{ entry().date }}</span>
    </mat-card-subtitle>
    <div class="header-icons">
      <span [matMenuTriggerFor]="menu">
        <mat-icon>more_vert</mat-icon>
      </span>
    </div>
  </mat-card-header>
  <mat-card-content class="entry-content px-3">
    <p class="entry-description mb-4">
      {{
        entry().description ? entry().description : "No description provided."
      }}
    </p>
    <hr />
  </mat-card-content>
  <mat-card-actions
    class="d-flex justify-content-end ps-2 pe-3 pt-1 entry-actions"
  >
    <span class="push-right">ID: {{ entry().id }}</span>
  </mat-card-actions>
</mat-card>

<mat-menu #menu="matMenu">
  <button mat-menu-item (click)="openDialog()">
    <mat-icon>delete</mat-icon>
    <span>Delete Entry</span>
  </button>
</mat-menu>
