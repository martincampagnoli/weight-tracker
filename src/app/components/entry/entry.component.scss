.entry-card {
  border-radius: 12px;
  background-color: #ffffff;
}

.entry-header {
  display: flex;
  align-items: center;
  padding-bottom: 5px;
  padding-top: 5px !important;
  background-color: #f7f9ff;
  border-bottom: 1px solid rgba(121, 116, 126, 0.12);
  border-top-left-radius: 12px;
  border-top-right-radius: 12px;
}

.entry-title {
  font-size: large;
}

.entry-subtitle {
  font-size: medium;
}

.entry-content {
  padding-top: 16px;
}

.entry-description {
  font-size: small;
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

.entry-actions {
  font-size: small;
}

.header-icons {
  display: flex;
  align-items: center;
  margin-left: auto;
}

mat-icon {
  cursor: pointer;
  margin-left: 8px;
}
