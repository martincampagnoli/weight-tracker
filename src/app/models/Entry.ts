export interface Entry {
  id: number;
  weight: number;
  date: string;
  description?: string;
}
