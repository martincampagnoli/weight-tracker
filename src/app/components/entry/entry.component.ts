import { Component, DestroyRef, inject, input } from '@angular/core';
import { MatDialog } from '@angular/material/dialog';
import {
  MatCard,
  MatCardActions,
  MatCardContent,
  MatCardHeader,
  MatCardSubtitle,
  MatCardTitle,
} from '@angular/material/card';
import { MatIcon } from '@angular/material/icon';
import { MatMenu, MatMenuItem, MatMenuTrigger } from '@angular/material/menu';
import { CommonModule } from '@angular/common';
import { MatFormFieldModule } from '@angular/material/form-field';
import { ConfirmDeleteDialogComponent } from '../confirm-delete-dialog/confirm-delete-dialog.component';
import { Store } from '@ngrx/store';
import { deleteEntrySuccess } from 'src/app/store/default';
import { SnackBarUtil } from 'src/app/utils/snackbar.util';
import { Entry } from 'src/app/models/Entry';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';

@Component({
  selector: 'app-entry',
  templateUrl: './entry.component.html',
  styleUrls: ['./entry.component.scss'],
  standalone: true,
  imports: [
    MatCard,
    MatCardTitle,
    MatCardContent,
    MatCardActions,
    MatIcon,
    MatCardSubtitle,
    MatCardHeader,
    MatMenu,
    MatMenuTrigger,
    MatMenuItem,
    CommonModule,
    MatFormFieldModule,
  ],
})
export class EntryComponent {
  private store: Store = inject(Store);
  private snackBarUtil: SnackBarUtil = inject(SnackBarUtil);
  private destroyRef = inject(DestroyRef);
  readonly entry = input.required<Entry>();
  readonly dialog = inject(MatDialog);

  deleteEntry(): void {
    this.store.dispatch(
      deleteEntrySuccess({ payload: { id: this.entry().id } })
    );
    this.snackBarUtil.show('Entry deleted successfully');
  }

  openDialog(): void {
    const dialogRef = this.dialog.open(ConfirmDeleteDialogComponent, {
      data: { entryId: this.entry().id },
      width: '250px',
      enterAnimationDuration: '0ms',
      exitAnimationDuration: '0ms',
    });

    dialogRef
      .afterClosed()
      .pipe(takeUntilDestroyed(this.destroyRef))
      .subscribe((confirm) => {
        if (confirm) {
          this.deleteEntry();
        }
      });
  }
}
