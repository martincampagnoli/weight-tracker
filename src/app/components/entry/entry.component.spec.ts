import { ComponentFixture, TestBed } from '@angular/core/testing';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { provideStore } from '@ngrx/store';

import { EntryComponent } from './entry.component';
import { Entry } from 'src/app/models/Entry';

describe('EntryComponent', () => {
  let component: EntryComponent;
  let fixture: ComponentFixture<EntryComponent>;

  const mockEntry: Entry = {
    id: 1,
    weight: 70.5,
    date: '01-01-2024',
    description: 'Test entry'
  };

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        EntryComponent,
        NoopAnimationsModule
      ],
      providers: [
        provideStore()
      ]
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(EntryComponent);
    component = fixture.componentInstance;
    
    // Set required input
    fixture.componentRef.setInput('entry', mockEntry);
    fixture.detectChanges();
  });

  describe('Component Creation', () => {
    it('should create', () => {
      expect(component).toBeTruthy();
    });

    it('should match snapshot', () => {
      expect(fixture).toMatchSnapshot();
    });
  });

  describe('Input Properties', () => {
    it('should display entry data correctly', () => {
      const compiled = fixture.nativeElement;
      expect(compiled.textContent).toContain('70.5');
      expect(compiled.textContent).toContain('01-01-2024');
      expect(compiled.textContent).toContain('Test entry');
    });
  });
});
