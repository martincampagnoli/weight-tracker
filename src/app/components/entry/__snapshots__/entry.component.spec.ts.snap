// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntryComponent Component Creation should match snapshot 1`] = `
<app-entry
  destroyRef={[Function NodeInjectorDestroyRef]}
  dialog={[Function _MatDialog]}
  entry={[Function Function]}
  snackBarUtil={[Function SnackBarUtil]}
  store={[Function _Store]}
>
  <mat-card
    appearance="outlined"
    class="mat-mdc-card mdc-card entry-card mat-mdc-card-outlined mdc-card--outlined"
  >
    <mat-card-header
      class="mat-mdc-card-header entry-header"
    >
      <div
        class="mat-mdc-card-header-text"
      >
        <mat-card-title
          class="mat-mdc-card-title entry-title"
        >
          <span>
            70.5 kgs
          </span>
        </mat-card-title>
        <mat-card-subtitle
          class="mat-mdc-card-subtitle entry-subtitle"
        >
          <span>
            01-01-2024
          </span>
        </mat-card-subtitle>
      </div>
      <div
        class="header-icons"
      >
        <span
          aria-expanded="false"
          aria-haspopup="menu"
          class="mat-mdc-menu-trigger"
        >
          <mat-icon
            aria-hidden="true"
            class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
            data-mat-icon-type="font"
            role="img"
          >
            more_vert
          </mat-icon>
        </span>
      </div>
    </mat-card-header>
    <mat-card-content
      class="mat-mdc-card-content entry-content px-3"
    >
      <p
        class="entry-description mb-4"
      >
         Test entry 
      </p>
      <hr />
    </mat-card-content>
    <mat-card-actions
      class="mat-mdc-card-actions mdc-card__actions d-flex justify-content-end ps-2 pe-3 pt-1 entry-actions"
    >
      <span
        class="push-right"
      >
        ID: 1
      </span>
    </mat-card-actions>
  </mat-card><mat-menu>
  </mat-menu>
</app-entry>
`;
